### DISCUSSION TITLE
Pocketbase + stripe

### DISCUSSION THREAD

**Discussion by AndrewDownsie22** [pocketbase/pocketbase#1454] [Q&A]
Starting a conversation here.... looking to use pocketbase with a small web application in nextjs. Has anyone successfully paired it with stripe or another payment processing API?

**Comment by R1ky68** [comment]
Hi! Like @ganigeorgiev answered in [this post](https://github.com/pocketbase/pocketbase/discussions/1290) one of the ways to pair stripe with PocketBase is by [using PocketBase as framework](https://pocketbase.io/docs/use-as-framework/), in that way, you'll be able to have your own logic by adding new routes to your backend. Here are the steps:
### 1. Create a go project with a new pocketbase route in it
main.go

    func main() {
	  app := pocketbase.New()
  
	  app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		  e.Router.AddRoute(echo.Route{
			  Method:  http.MethodPost,
			  Path:    "/api/create-payment-intent",
			  Handler: CreatePaymentIntentHandler,
		  })
		  return nil
	  })
  
	  if err := app.Start(); err != nil {
		  log.Fatal(err)
	  }
    }

The scope of this route is to return a **clientSecret** string, in order to initialize the [Stripe Payment Element](https://stripe.com/docs/payments/payment-element)

### 2. Create the handler _CreatePaymentIntentHandler_
main.go

    func CreatePaymentIntentHandler(c echo.Context) error {
	  clientSecret, err := stripeInstance.CreatePaymentIntent()
	  if err != nil {
		  apis.NewApiError(http.StatusInternalServerError, "Error", err)
		  return err
	  }
  
	  return c.JSON(http.StatusOK, clientSecret)
    }

stripe/stripe.go

    func CreatePaymentIntent() (clientSecret string, err error) {
	  stripe.Key = "YOUR_SECRET_KEY"
  
	  // Create a PaymentIntent with amount and currency
	  params := &stripe.PaymentIntentParams{
		  PaymentMethod: stripe.String("pm_card_visa"),
		  Amount:        stripe.Int64(YOUR_AMOUNT),
		  Currency:      stripe.String(string(stripe.CurrencyYOUR_CURRENCY)),
		  AutomaticPaymentMethods: &stripe.PaymentIntentAutomaticPaymentMethodsParams{
			  Enabled: stripe.Bool(true),
		  },
	  }
  
	  pi, err := paymentintent.New(params)
	  if err != nil {
		  return "", err
	  }
  
	  return pi.ClientSecret, nil
    }

Note that I can call the _CreatePaymentIntent_ function by this import:
`stripeInstance "github.com/YOUR_USERNAME/YOUR_REPOSITORY/stripe"`

### 3. Prepare the frontend 
This is an example using Nuxt 3, but it's just a small adaptation of the HTML & JavaScript code used in the [Custom payment flow](https://stripe.com/docs/payments/quickstart)

      <template>
      <div class="flex justify-center items-center">
          <form id="payment-form" @submit="handleSubmit">
              <div id="payment-element"></div>
              <button id="submit">
                  <div class="spinner hidden" id="spinner"></div>
                  <span id="button-text">Pay now</span>
              </button>
              <div id="payment-message" class="hidden"></div>
          </form>
      </div>
      </template>
  
      <script setup>
	  import { loadStripe } from '@stripe/stripe-js/pure'
	  import PocketBase from 'pocketbase';
	  import { BASE_URL, PUBLISHABLE_KEY_STRIPE } from '~~/utils/constants';
  
	  let stripe, elements, clientSecret;
  
	  watchEffect(async () => {
		  await initialize();
		  checkStatus();
	  })
  
	  // Fetches a payment intent and captures the client secret
	  async function initialize() {
		  stripe = await loadStripe(PUBLISHABLE_KEY_STRIPE);
		  const pb = new PocketBase(BASE_URL);
  
		  clientSecret = await pb.send("/api/create-payment-intent", {
			  method: "POST",
		  });
  
		  const appearance = {
			  theme: 'stripe'
		  };
		  elements = stripe.elements({ appearance, clientSecret });
  
		  const paymentElementOptions = {
			  layout: "tabs",
		  };
  
		  const paymentElement = elements.create("payment", paymentElementOptions);
		  paymentElement.mount("#payment-element");
	  }
  
	  async function handleSubmit(e) {
		  e.preventDefault();
		  setLoading(true);
  
		  const { paymentIntent, error } = await stripe.confirmCardPayment(clientSecret);
		  handlePayment(paymentIntent, error);
  
		  setLoading(false);
	  }
  
	  // Fetches the payment intent status after payment submission
	  async function checkStatus() {
		  if (!clientSecret) {
		  console.log("No client secret found");
		  return;
		  }
  
		  const { paymentIntent } = await stripe.retrievePaymentIntent(clientSecret);
  
		  switch (paymentIntent.status) {
		  case "succeeded":
			  showMessage("Payment succeeded!");
			  break;
		  case "processing":
			  showMessage("Your payment is processing.");
			  break;
		  case "requires_payment_method":
			  showMessage("Your payment was not successful, please try again.");
			  break;
		  default:
			  showMessage("Something went wrong.");
			  break;
		  }
	  }
  
	  // ------- UI helpers -------
  
	  function showMessage(messageText) {
		  const messageContainer = document.querySelector("#payment-message");
  
		  messageContainer.classList.remove("hidden");
		  messageContainer.textContent = messageText;
  
		  setTimeout(function () {
		  messageContainer.classList.add("hidden");
		  messageText.textContent = "";
		  }, 4000);
	  }
  
	  // Show a spinner on payment submission
	  function setLoading(isLoading) {
		  if (isLoading) {
		  // Disable the button and show a spinner
		  document.querySelector("#submit").disabled = true;
		  document.querySelector("#spinner").classList.remove("hidden");
		  document.querySelector("#button-text").classList.add("hidden");
		  return
		  } 
  
		  document.querySelector("#submit").disabled = false;
		  document.querySelector("#spinner").classList.add("hidden");
		  document.querySelector("#button-text").classList.remove("hidden");
	  }
  
	  function handlePayment(paymentIntent, error) {
		  if (error || !paymentIntent || paymentIntent.status !== 'succeeded') {
			  console.log("2", error);
			  return
		  }
                  
                  // Here handle the paymentSucceeded such as going to the home page
	  }
      </script>
  
      <style scoped>
          /* Variables */
          * {
             box-sizing: border-box;
             }
      
      body {
        font-family: -apple-system, BlinkMacSystemFont, sans-serif;
        font-size: 16px;
        -webkit-font-smoothing: antialiased;
        display: flex;
        justify-content: center;
        align-content: center;
        height: 100vh;
        width: 100vw;
      }
      
      form {
        width: 30vw;
        min-width: 500px;
        align-self: center;
        box-shadow: 0px 0px 0px 0.5px rgba(50, 50, 93, 0.1),
          0px 2px 5px 0px rgba(50, 50, 93, 0.1), 0px 1px 1.5px 0px rgba(0, 0, 0, 0.07);
        border-radius: 7px;
        padding: 40px;
      }
      
      .hidden {
        display: none;
      }
      
      #payment-message {
        color: rgb(105, 115, 134);
        font-size: 16px;
        line-height: 20px;
        padding-top: 12px;
        text-align: center;
      }
      
      #payment-element {
        margin-bottom: 24px;
      }
      
      /* Buttons and links */
      button {
        background: #5469d4;
        font-family: Arial, sans-serif;
        color: #ffffff;
        border-radius: 4px;
        border: 0;
        padding: 12px 16px;
        font-size: 16px;
        font-weight: 600;
        cursor: pointer;
        display: block;
        transition: all 0.2s ease;
        box-shadow: 0px 4px 5.5px 0px rgba(0, 0, 0, 0.07);
        width: 100%;
      }
      button:hover {
        filter: contrast(115%);
      }
      button:disabled {
        opacity: 0.5;
        cursor: default;
      }
      
      /* spinner/processing state, errors */
      .spinner,
      .spinner:before,
      .spinner:after {
        border-radius: 50%;
      }
      .spinner {
        color: #ffffff;
        font-size: 22px;
        text-indent: -99999px;
        margin: 0px auto;
        position: relative;
        width: 20px;
        height: 20px;
        box-shadow: inset 0 0 0 2px;
        -webkit-transform: translateZ(0);
        -ms-transform: translateZ(0);
        transform: translateZ(0);
      }
      .spinner:before,
      .spinner:after {
        position: absolute;
        content: "";
      }
      .spinner:before {
        width: 10.4px;
        height: 20.4px;
        background: #5469d4;
        border-radius: 20.4px 0 0 20.4px;
        top: -0.2px;
        left: -0.2px;
        -webkit-transform-origin: 10.4px 10.2px;
        transform-origin: 10.4px 10.2px;
        -webkit-animation: loading 2s infinite ease 1.5s;
        animation: loading 2s infinite ease 1.5s;
      }
      .spinner:after {
        width: 10.4px;
        height: 10.2px;
        background: #5469d4;
        border-radius: 0 10.2px 10.2px 0;
        top: -0.1px;
        left: 10.2px;
        -webkit-transform-origin: 0px 10.2px;
        transform-origin: 0px 10.2px;
        -webkit-animation: loading 2s infinite ease;
        animation: loading 2s infinite ease;
      }
      
      @-webkit-keyframes loading {
        0% {
          -webkit-transform: rotate(0deg);
          transform: rotate(0deg);
        }
        100% {
          -webkit-transform: rotate(360deg);
          transform: rotate(360deg);
        }
      }
      @keyframes loading {
        0% {
          -webkit-transform: rotate(0deg);
          transform: rotate(0deg);
        }
        100% {
          -webkit-transform: rotate(360deg);
          transform: rotate(360deg);
        }
      }
      
      @media only screen and (max-width: 600px) {
        form {
          width: 80vw;
          min-width: initial;
        }
      }
      </style>

Here the most important stuff is the _initialize_ function, which makes an API call to our new PocketBase route, saving the string returned from the call to the variable _clientSecret_ (it doesn't need to be a POST request); and the _handleSubmit_ function. 

I hope this will be useful to someone and that the process is clear.
